### ISSUE TITLE
emit event once initialised

### ISSUE THREAD

**Issue by pgte** [ipfs/js-ipfs#1058]
- **Version**: 0.26.0
- **Platform**: All
- **Subsystem**:

#### Type:

Enhancement

#### Severity:

Very Low

#### Description:

I want to support better offline use cases. I need the peerId to be able to start working locally, without needing for Swarm, Bitswap or Pubsub to consider themselves ready.

AFAIK, right now there is no event that tells me this. `init` is only emitted if the repo needs initialising.
`start` and `ready` are only emitted once the network layer is ready, which may not happen in my case.

Right now I'm polling the IPFS node for the `_peerInfo` property, but I would like to have a better alternative.

I would like to have an event that tells me once the peerId is known.

#### Steps to reproduce the error:

N/A

**Comment by VictorBjelkholm** [comment]
> AFAIK, right now there is no event that tells me this. init is only emitted if the repo needs initialising.
start and ready are only emitted once the network layer is ready, which may not happen in my case.

This is incorrect regarding the `init` event. It's supposed to be emitted when the node is done initializing, not if it needs initializing. Things seem to work currently, check this example:

```
const ipfs = require('./src')

const node = new ipfs({repo: './repo-' + Math.random(), init: {bits: 32}, start: false})
node.once('ready', () => {
  console.log('ready')
  console.log(`current state is "${node.state._state}"`)
})
node.once('start', () => {
  console.log('started but it should have had started')
})
node.once('init', () => {
  console.log('initialized')
  console.log(`current state is "${node.state._state}"`)
})
```

This outputs the following for me:

```
$ node node-init-test.js
initialized
current state is "stopped"
ready
current state is "stopped"
```

However, I would expect `node._peerInfo` to actually have the peerInfo at this point, but it's `undefined`. So calling `node.id()` would not work.

**Comment by VictorBjelkholm** [comment]
Ah, I think I misunderstood a tiny bit. You're saying that if the repo is already initiralized, the `init` event is never emitted, correct? That seems to be true, while the `ready` event is still emitted in that case.

**Comment by pgte** [comment]
@VictorBjelkholm yes, that's the problem. If the node happens to be offline, the `ready` event is not emitted, and I need a peerId in order to be able to work locally...

**Comment by wraithgar** [comment]
Made a PR to address this that goes down the path of correcting the `init` event behavior to emit when the node is done initializing, not only just when it has initialized a new repo.

**Comment by Mr0grog** [comment]
@pgte I just spent a bunch of time staring at all the bootstrapping code to document the constructor options and it looks to me like `ready` should *always* emit (so long as there were no errors). I just took my computer offline and tried this; it fired the `ready` event (and got valid peer info):

```js
var IPFS = require('ipfs');
var node = new IPFS();
node.on('ready', () => {
  console.log('READY!!!!!');
  node.id().then(id => console.log('Peer Info:', id));
});
```

Am I misunderstanding what you’re getting at? (Seems important since these [new event docs](https://github.com/ipfs/js-ipfs/pull/1294/files#diff-04c6e90faac2675aa89e2176d2eec7d8R256) would be wrong if there’s a non-error case where it doesn’t emit.)

**Comment by pgte** [comment]
I now see that I should have added a test case, since the behaviour I can no longer reproduce the behaviour I think was witnessing.

(I think this may related to the `ws-star` transport. It still doesn't allow me to start offline if I enable this transport by declaring a `p2p-websocket-star` address. — but this is a completely separate issue).

Anyway, the point here is:

The `ready` event (if the `ws-star` transport was able to start offline) is not really what I need, since it is emitted after the network layer is initialised (which is later than I need).
I would love to have an event which tells me once I have a PeerId and functioning repo so I can start making local changes, being it `init` or some other.

**Comment by alanshaw** [comment]
Would passing start:false to the constructor and then waiting for the ready event be what you’re looking for?
